### PR TITLE
Add IR verification.

### PR DESCRIPTION
**Pull request by otrho**
When building a new IR context one can now call `.verify()` and it'll essentially lint it.

There are a bunch of new verification errors which are mostly fairly terse and could be made richer in the future.  Also, the reporting of these errors can be improved by providing a string with the offending value, much like the spans in core.  This isn't the highest priority right now though.

I found a ~day~ week or two's worth of bugs which also needed to be fixed, mostly around type mismatches.

And then I implemented `if let` for IR, which involved updating how structs are handled and local symbol shadowing and also needed an integer `cmp` instruction in IR to compare tags.

### CODE REVIEW

**Comment by otrho** [comment]
I've been checking the E2E tests and I seem to have 'broken' a bunch of them (when using IR) with this PR.  'Broken' generally means verification has found a bunch more bugs which I didn't expect, since all the unit tests are passing.

File: sway-ir/src/verify.rs [hunk 1]
```diff
@@ -69,220 +83,450 @@ impl Context {
...
+    //        &self.context.values[ptr_val.0].value,
+    //        ValueDatum::Instruction(Instruction::GetPointer(_))
+    //            | ValueDatum::Instruction(Instruction::PointerCast(..))
+    //    ) {
+    //        Ok(())
+    //    } else {
+    //        Err(IrError::VerifyPtrCastFromNonPointer)
+    //    }
+    //}
+
```
Comment (sezna) [inline sway-ir/src/verify.rs:433]: We should delete these if they're not being used 
```suggestion
```

**Review by sezna — APPROVED** [review]

**Comment by mohammadfawaz** [comment]
Will you please add some IR tests that use the new `cmp` instruction? Can probably start with a simple code that uses `if let` and go from there.

**Comment by otrho** [comment]
> Will you please add some IR tests that use the new `cmp` instruction? Can probably start with a simple code that uses `if let` and go from there.

Ooh, good call.  I've added one and found that I forgot to add `cmp` to the parser!

**Review by mohammadfawaz — APPROVED** [review]

File: sway-core/src/asm_generation/from_ir.rs [hunk 2]
```diff
@@ -1341,46 +1331,38 @@ impl<'ir> AsmBuilder<'ir> {
...
-                    "is" => ConstantRegister::InstructionStart,
-                    "ret" => ConstantRegister::ReturnValue,
-                    "retl" => ConstantRegister::ReturnLength,
-                    "flag" => ConstantRegister::Flags,
-                    _ => unreachable!(
-                        "Unrecognized special register. This should have been \
-                                          caught in the verifier"
-                    ),
+                VirtualRegister::Constant(match reg {
+                    sway_ir::Register::Of => ConstantRegister::Overflow,
```
Comment (mohammadfawaz) [inline sway-core/src/asm_generation/from_ir.rs:1340]: Nice. Thanks!

**Review by mohammadfawaz — APPROVED** [review]
LGTM! Thanks for cleaning up the contract call stuff.

**Review by sezna — APPROVED** [review]

**Pull request merged**